### ISSUE TITLE
Set Z in resigned training games to e.g. +/- 0.95 instead of +/- 1.00

### ISSUE THREAD

**Issue by killerducky** [LeelaChessZero/lc0#768]
Currently game results (Z) for resigned games are stored as +1.00 or -1.00. But some of those games are actually draws or losses instead of wins. We have resign play through to correct this, but I propose for the resigned games we tell the NN there is some uncertainty in the result of the game.

A few different methods to pick the value: 

1) Based on the Q of the game at the point of resign. Ex: ResignPercentage is set to 0.95, but Q of the game suddenly overshoots this to 0.98. That side resigns, and Z is set to 0.98.

2) Use observed score of games played with ResignPlaythrough that also *would* have been resigned by 
one side. Ex: after a game would have been won by resign, that player scores 0.95 points (some combination of 5% losses or 2.5% draws). Set Z to 0.95. Server would measure this value (it already measures something similar) and either pass to clients, or modify training data on server side.

I heard about this idea from @lightvector. Maybe this would help prevent the NN getting saturated with win rates too close to 1.0.

**Comment by killerducky** [comment]
lightvector has released a paper on his methods: https://arxiv.org/abs/1902.10565

6.2.1 Reducing Playouts Instead of Resignation

Unlike AlphaZero or Leela Zero, KataGo does not terminate games early via resignation during self-play.

Instead, in the case of extreme winning chances, KataGo reduces the number of visits used in the search. During self-play, if both sides agree that for the last 5 turns, the worst MCTS winrate estimate p for the losing side has been less than 5%, then the number of visits is capped to λn + (1 − λ)N where n and N are the small and large limits used in visit cap oscillation and λ = p/0.05 is the proportion of the way that p is from 5% to 0%. Additionally, the training positions are recorded with only 0.1 + 0.9 ∗ λ weight, downweighting training samples where the original AlphaZero process would have already resigned.

Playing out the full game allows the final determination of the ownership of all areas and the final score difference for training the auxiliary targets. Additionally, avoiding resignation also reduces the number of positions that are incorrectly recorded, improving the quality of the data.

**Comment by Tilps** [comment]
Saturating the win rates is part of the point of resign - providing an upward pressure to offset some of the noisiness of MCTS.  The playthrough itself provides the averaging to reduce that pressure a bit.  Removing the pressure entirely doesn't seem ideal.
However Q learning could also be used to reduce the pressure a bit more.

**Comment by lightvector** [comment]
It feels like "wanting" to saturate win rates is a strange goal. If that's what you wanted, you could always artificially do so by simply applying an additional penalty term in the loss function that penalizes it for being close to 0.5 and drives it symmetrically away towards 0 and 1 with some gradient (regardless of which of 0 or 1 was actually correct in a position). But I'd be a little surprised if that turned out to be a good idea.

I'd think you only "want" to saturate the winrates when you can do so in a way that makes the neural net be *more correct*, rather than in any arbitrary way. So it seems to me that the purpose of resignation is to speed up training and to reduce the number of training samples from positions that are less important (because the game is likely decided and uninteresting for the neural net to train as much on as on more complex positions where good evaluation matters more), not because you actually "want" the value sharpness. The small bit of additional value sharpness you get (scoring a small percent of games that actually would have been wins as losses, and vice versa) seems like the kind that, like the silly penalty term in the previous paragraph, would push the neural net towards being more confident without actually being more correct, and is a minor unintended side-effect rather than a goal.

**Comment by Tilps** [comment]
As I said, saturating the win rates through resign is intended to offset some of the noisiness of MCTS - which is indeed about making it more correct.  Resign threshold should be selected such that the number of incorrect resigns is significantly less as a percentage, than the resign threshold itself.  If that stops being true you need to reduce that pressure. (Unfortunately its pretty impractical to test such a statement in our current configuration - so it could be that our resign threshold is sometimes set too aggressively.)
With the saturation making things more correct, that also cascades forwards into earlier resigns, allowing even more focus where its most useful.

**Comment by lightvector** [comment]
@Tilps - just curious - what noisiness are you referring to such that it should be an improvement to set the training data to make the neural net value more extreme?

* Is it that _when the raw neural net reports a value like 0.95_, you're seeing that the empirical mean win/loss/draw outcome among such positions in training is actually systematically greater than 0.95? In this case, yes, doing something to make the neural net underfit less should be desirable, although ideally you'd want to do that by finding and addressing the fundamental causes of why it's underfitting in training.

* Is it that _when MCTS reports a value like 0.95_, you're seeing that the empirical mean win/loss/draw outcome among such positions in training is actually systematically greater than 0.95? If that's so, I wouldn't think that's a problem at all. An MCTS search saying 0.95 at the root isn't supposed to be a claim in the first place that the mean outcome from the position is 0.95, rather it's literally that the *average position in the search* is 0.95, which since it's aggregating over lots of possible variations gives evidence that the root position is actually *better* than 0.95. 
So it wouldn't mean that the neural net is wrong or underconfident on any individual position within the search for saying 0.95 and that the neural net needs to be made more confident. Rather, it's a simple reflection of the fact that the search has access to a lot more information any single neural net evaluation does - depending on the correlation structure, an entire subtree averaging that close to winning, where the search in particular has *not* found any scary variations that would have dropped the value a lot, can easily be much stronger evidence than a neural net's guess on a single position.

* Is it that you concerned about that MCTS self-play will sometimes blunder positions, and that's what you mean by "noise", and that resigning is a way of making sure the "proper" side wins before having a chance to blunder? I'm guessing this probably isn't what you meant, since of course with resignation the winning side is also going to be winning games that properly it shouldn't have (since the position was *actually* losing and the neural net was mistaken about some aspect of that position). At least, resignation isn't going to be any better at estimating the "true" value of the position than playing the position out, probably actually worse.

Or not quite any of the above?

@killerducky your number 2 item is what I would lean towards doing, because of this point about the MCTS winrate potentially not being the expected outcome. To the degree that they do diverge (i.e. a resign threshold of X actually produces an empirical error rate << X),  you prefer to train the neural net towards your actual best guess of the value of the game, and in that case your actual best guess would be the empirical error rate, not the MCTS value.

Or at least, a starting point of intuition should be that if you want a model to make a best guess of the truth, then you should make your training data your best guess of truth that you can afford. (A second-order thing would be that depending on how much it actually cumulatively changes value sharpness, you might need to re-tune c_PUCT, since c_PUCT is what controls the relative weighting of value versus policy, and if you're changing value scaling you might need to change that constant).

**Comment by Tilps** [comment]
My claim is that when the average W value of the best child in search is greater than 0.96 - that in games played from that position under non-training conditions the actual frequency of win is much greater than 0.96 - thus training the net towards 1.0 is a more correct reward than training it towards 0.96.

**Comment by lightvector** [comment]
Yep, absolutely. Lets' say that when you use 0.96 as a threshold the actual empirical frequency of win is such that the average real reward under those conditions is 0.99. Then in in that case, 1.0 is a more correct reward than than 0.96, but 0.99 is an even more correct reward than 1.0, which is what I believe killerducky's number 2 proposal amounts to.

This is also exactly my second bullet point above. And note critically that while 0.99 would be the theoretical correct reward, which is more than 0.96, it does *not* necessarily mean that the neural net is incorrectly being underconfident.

**Comment by Tilps** [comment]
The 20% playthrough already means the average target is ~0.99 instead of 1.0.  Then if we add in Q learning the average target may become 0.98 instead.
Actually building a framework to accurately measure the 'true' reward target is hard.  Building an approximation of that instead is what will probably happen - and then will that approximation be better than the current status relative to the true reward target?  The proposal to use resign playthrough (option 2) is in my mind likely to be a worse approximation than the current one of using 80% unity and 20% playthrough.  Using pure Q also worse approximation.  Using average of 10% playthrough and 40% unity and 50% Q (that would come from using (Q+Z)/2) - maybe better... I'm not sure.

**Comment by jhellis3** [comment]
Part of the problem is play-through can actually damage the target due to temperature induced blunders.  Probably far more damaging than 0.9x target since you are decaying to draw or even full result flip even if only a small portion of the time. 

(Q+Z) / 2 works because move selection and scores only matter relative to everything else which is based upon the same criteria.  Q alone is the perfect training criteria if you know for certain the NN is already optimal or "perfect" for its size.  But of course, we don't know that (and it isn't) so (Q + Z) / 2 makes sense.  (Q + Z) / 2 also allows the NN to not only quantify and distinguish the score of a node but also how easy or certain that score is to attain.

**Comment by Naphthalin** [comment]
As the current behavior seems to be intended, and false resign rate is tracked, do we need this issue to be open or can it be closed?
